### PR TITLE
PERF: assert_frame_equal / assert_series_equal

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.2.0.rst` file if fixing a bug or adding a new feature.

Follow-up to #55949 addressing some of @jbrockmendel's comments and also expanding to non-multiindex cases:

```
import pandas as pd

N = 1_000_000

idx = pd._testing.makeStringIndex(N)

ser1 = pd.Series(1, index=idx.copy(deep=True))
ser2 = pd.Series(1, index=idx.copy(deep=True))

%timeit pd.testing.assert_series_equal(ser1, ser2)

# 17.8 s ± 129 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# 48.1 ms ± 3.28 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

### CODE REVIEW

File: pandas/_testing/asserters.py [hunk 1]
```diff
@@ -328,7 +319,7 @@ def _get_ilevel_values(index, level):
             diff = np.sum(mismatch.astype(int)) * 100.0 / len(left)
             msg = f"{obj} values are different ({np.round(diff, 5)} %)"
             raise_assert_detail(obj, msg, left, right)
-    else:
+    elif not left.equals(right):
```
Comment (jbrockmendel) [inline pandas/_testing/asserters.py:322]: do we know that .equals is the correct amount of strict?

Comment (lukemanley) [reply pandas/_testing/asserters.py]: I think so, but after reviewing again I've reverted this line in favor of a lower level change to `_array_equivalent_object`. 

This has the benefit of improving performance for more cases, e.g. not just the index values but in this case the series values:

```
import pandas as pd

N = 100_000

values = pd._testing.makeStringIndex(N).values

ser1 = pd.Series(values.copy())
ser2 = pd.Series(values.copy())

%timeit pd.testing.assert_series_equal(ser1, ser2)

1.92 s ± 56.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
22.3 ms ± 4.18 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

File: pandas/core/dtypes/missing.py [hunk 2]
```diff
@@ -567,7 +567,22 @@ def _array_equivalent_object(left: np.ndarray, right: np.ndarray, strict_nan: bo
 
         return lib.array_equivalent_object(ensure_object(left), ensure_object(right))
 
-    for left_value, right_value in zip(left, right):
+    mask = isna(left) & isna(right)
+    try:
+        if not lib.array_equivalent_object(
+            ensure_object(left[~mask]),
```
Comment (mroeschke) [inline pandas/core/dtypes/missing.py:573]: Is it worth skipping the masking if there are no nulls in left and right?

Comment (lukemanley) [reply pandas/core/dtypes/missing.py:573]: good point. updated

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lukemanley

File: pandas/core/dtypes/missing.py [hunk 3]
```diff
@@ -562,12 +562,29 @@ def _array_equivalent_datetimelike(left: np.ndarray, right: np.ndarray):
 
 
 def _array_equivalent_object(left: np.ndarray, right: np.ndarray, strict_nan: bool):
-    if not strict_nan:
-        # isna considers NaN and None to be equivalent.
-
-        return lib.array_equivalent_object(ensure_object(left), ensure_object(right))
-
-    for left_value, right_value in zip(left, right):
+    left = ensure_object(left)
```
Comment (jbrockmendel) [inline pandas/core/dtypes/missing.py:565]: can we get here in cases that are not already object?  seems like if that happens then something has gone wrong?

**Pull request merged**